op-supervisor/supervisor/backend/db/fromda/db.go

```diff
@@ -0,0 +1,235 @@
+}
+
+// LatestDerivedFrom returns the last known primary key (the L1 block)
+func (db *DB) LatestDerivedFrom() (ref types.BlockSeal, err error) {
+	db.rwLock.RLock()
+	defer db.rwLock.RUnlock()
+	lastIndex := db.store.LastEntryIdx()
+	if lastIndex < 0 {
+		return types.BlockSeal{}, entrydb.ErrFuture
+	}
+	last, err := db.readAt(lastIndex)
+	if err != nil {
+		return types.BlockSeal{}, fmt.Errorf("failed to read last derived-from data: %w", err)
+	}
+	return last.derivedFrom, nil
+}
+
+// LatestDerived returns the last known value (the L2 block that was derived)
+func (db *DB) LatestDerived() (ref types.BlockSeal, err error) {
+	db.rwLock.RLock()
+	defer db.rwLock.RUnlock()
+	lastIndex := db.store.LastEntryIdx()
+	if lastIndex < 0 {
+		return types.BlockSeal{}, entrydb.ErrFuture
+	}
+	last, err := db.readAt(lastIndex)
+	if err != nil {
+		return types.BlockSeal{}, fmt.Errorf("failed to read last derived data: %w", err)
+	}
+	return last.derived, nil
+}
+
+// LastDerivedAt returns the last L2 block derived from the given L1 block.
+func (db *DB) LastDerivedAt(derivedFrom eth.BlockID) (derived types.BlockSeal, err error) {
+	db.rwLock.RLock()
+	defer db.rwLock.RUnlock()
+	_, link, err := db.lastDerivedAt(derivedFrom.Number)
+	if err != nil {
+		return types.BlockSeal{}, err
+	}
+	if link.derivedFrom.ID() != derivedFrom {
+		return types.BlockSeal{}, fmt.Errorf("searched for last derived-from %s but found %s: %w",
+			derivedFrom, link.derivedFrom, entrydb.ErrConflict)
+	}
+	return link.derived, nil
+}
+
+// DerivedFrom determines where a L2 block was first derived from.
+// (a L2 block may repeat if the following L1 blocks are empty and don't produce additional L2 blocks)
+func (db *DB) DerivedFrom(derived eth.BlockID) (derivedFrom types.BlockSeal, err error) {
+	db.rwLock.RLock()
+	defer db.rwLock.RUnlock()
+	_, link, err := db.firstDerivedFrom(derived.Number)
+	if err != nil {
+		return types.BlockSeal{}, err
+	}
+	if link.derived.ID() != derived {
+		return types.BlockSeal{}, fmt.Errorf("searched for first derived %s but found %s: %w",
+			derived, link.derived, entrydb.ErrConflict)
+	}
+	return link.derivedFrom, nil
+}
+
+func (db *DB) firstDerivedFrom(derived uint64) (entrydb.EntryIdx, LinkEntry, error) {
+	return db.find(false, func(link LinkEntry) int {
+		return cmp.Compare(link.derived.Number, derived)
+	})
+}
+
+func (db *DB) lastDerivedAt(derivedFrom uint64) (entrydb.EntryIdx, LinkEntry, error) {
+	// Reverse: prioritize the last entry.
+	return db.find(true, func(link LinkEntry) int {
+		return cmp.Compare(derivedFrom, link.derivedFrom.Number)
+	})
+}
+
+// find finds the first entry for which cmpFn(link) returns 0.
+// The cmpFn entries to the left should return -1, entries to the right 1.
+// If reverse, the cmpFn should be flipped too, and the last entry for which cmpFn(link) is 0 will be found.
+func (db *DB) find(reverse bool, cmpFn func(link LinkEntry) int) (entrydb.EntryIdx, LinkEntry, error) {
```

Comment:
Note to self (and others): The reason we can avoid using checkpoints in this structure is because we've established *all* entries to be a fixed width, and so we have a strongly defined row access. The implementation here anticipates that the fixed-width of the structures in the database won't ever change.

If the width of the data ever changes, *either* we would need to regenerate the whole database. And, if the data ever needs to become dynamic in size, checkpoints will need to be reintroduced.